delira/training/predictor.py

```diff
@@ -0,0 +1,286 @@
+
+        pred = self.module(
+            **mapped_data
+        )
+
+        # converts positional arguments and keyword arguments,
+        # but returns only keyword arguments, since positional
+        # arguments are not given.
+        return self._convert_to_npy_fn(
+            **pred
+        )[1]
+
+    def predict_data_mgr(self, datamgr, batchsize=None, metrics={},
+                         metric_keys=None, verbose=False):
+        """
+        Defines a routine to predict data obtained from a batchgenerator
+
+        Parameters
+        ----------
+        datamgr : :class:`BaseDataManager`
+            Manager producing a generator holding the batches
+        batchsize : int
+            Artificial batchsize (sampling will be done with batchsize
+            1 and sampled data will be stacked to match the artificial
+            batchsize)(default: None)
+        metrics : dict
+            the metrics to calculate
+        metric_keys : dict
+            the ``batch_dict`` items to use for metric calculation
+        verbose : bool
+            whether to show a progress-bar or not, default: False
+
+        Returns
+        -------
+        dict
+            a dictionary containing all predictions
+        dict
+            a dictionary containing all validation metrics (maybe empty)
+
+        """
+
+        orig_num_aug_processes = datamgr.n_process_augmentation
+        orig_batch_size = datamgr.batch_size
+
+        if batchsize is None:
+            batchsize = orig_batch_size
+
+        datamgr.batch_size = 1
+        datamgr.n_process_augmentation = 1
+
+        batchgen = datamgr.get_batchgen()
+
+        predictions_all, metric_vals = [], {k: [] for k in metrics.keys()}
+
+        n_batches = batchgen.generator.num_batches * batchgen.num_processes
+
+        if verbose:
+            iterable = tqdm(enumerate(batchgen), unit=' sample',
+                            total=n_batches, desc=self._tqdm_desc)
+
+        else:
+            iterable = enumerate(batchgen)
+
+        batch_list = []
+
+        for i, batch in iterable:
+
+            if not batch_list and (n_batches - i) < batchsize:
+                batchsize = n_batches - i
+                logger.debug("Set Batchsize down to %d to avoid cutting "
+                             "of the last batches" % batchsize)
+
+            batch_list.append(batch)
+
+            # if queue is full process queue:
+            if batchsize is None or len(batch_list) >= batchsize:
+
+                batch_dict = {}
+                for batch in batch_list:
+                    for key, val in batch.items():
+                        if key in batch_dict.keys():
+                            batch_dict[key].append(val)
+                        else:
+                            batch_dict[key] = [val]
+
+                for key, val_list in batch_dict.items():
+                    batch_dict[key] = np.concatenate(val_list)
+
+                preds = self.predict(batch_dict)
```

Comment:
do we allow

Reply:
you mean as input? This shouldn't matter, as long as your network/`prepare_batch` handle this correctly. 

For the output this becomes problematic, since we do some concatenation based on these values and we can't concatenate dicts. Another possibility would be to flatten these dicts by something like this:

```python
import collections

def flatten(d, parent_key='', sep='_'):
    items = []
    for k, v in d.items():
        new_key = parent_key + sep + k if parent_key else k
        if isinstance(v, collections.MutableMapping):
            items.extend(flatten(v, new_key, sep=sep).items())
        else:
            items.append((new_key, v))
    return dict(items)
```

This would be a design decision, we'd have to make now (as this currently causes the tests to fail). I don't really have an opinion on this. What do you think @ORippler @mibaumgartner @haarburger ?

Reply:
Yeah i also identified that as the reason for failure.

I dislike the idea of flattening the dict, as the new keys would need to be passed to `calc_metrics` and the user informed of this.

I would prefer a recursion-based solution to allow for nested dicts, though `calc_metrics` would also need to be updated (currently, the values used for metric calculation need to lie on the top-level of the dict).

Reply:
I'm not sure if it is beneficial to support nested dicts. Are there any useful scenarios where this could/must be used?

Reply:
I wouldn't like a recursion based approach. I can't see any benefits from it, but if we need to resolve arbitrary recursions multiple times this is bad style IMHO

Reply:
@mibaumgartner the reason we run into that failure is because [tf.Session.run()](https://www.tensorflow.org/api_docs/python/tf/Session#run) allows nested dicts and i used that functionality already inside `ClassificationNetworkBase` (it makes the logging of training loss easier as we have model.closure as static function for pytorch's multi-gpu approach). Could handle the unpacking inside the classes' `__call__` though.

Reply:
@ORippler I'm pro nested dict, if we can do this by resolving the recursion only once (i.e. something similar like flattening). What I don't like: If we would have to resolve the recursion mutiple times (i.e. one time for concatenation and another time for metric calculation (and maybe a third time for logging/reporting back to HP search in the future). This gets extremly inefficient quite fast.

Another problem would be, that we could have the same key in multiple subdicts and it is unclear which item the key refers to (e.g. in metric calculation).

Reply:
@mibaumgartner @ORippler I would like to move this on. What are we gonna do here?

Reply:
If it is benficial for tf applications, we should implement a soultion....

Reply:
But which one? The only one, I can think of, resolving the possible ambiguity in keys would be the one above. Alternatively we could wrap in in a `LookupConfig` (the parameters baseclass) and use `nested_get` here as well

Reply:
So, I just implemented the version with `LookupConfig` and nested dicts. There are only two issues left:

-  The `test_experiment_test_tf` functions seems to be unable to find the `dense_2` layer (which does not make sense to me)
-  `trixi` still has a problem with their dependencies, which is, why we had to downgrade the scipy version to 1.0.0, until they have released a package containing a fix for [this](https://github.com/MIC-DKFZ/trixi/issues/15). The problem is, that scipy 1.0.0 is not available for python 3.7. So for now the python 3.7 tests will be failing until the new trixi release.

I think, once the first issue is fixed, I will look to update all examples and documentation and then this is ready to merge; Any other opinions? @haarburger @ORippler @mibaumgartner ?